src/bn254/validation.rs

```diff
@@ -0,0 +1,137 @@
+}
+
+/// Checks that the passed point is on `BN256` curve.
+/// The `Infinity` point is not counted as on curve.
+pub(crate) fn is_on_curve<F: SmallField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    point: (&BN256BaseNNField<F>, &BN256BaseNNField<F>),
+    params: &Arc<BN256BaseNNFieldParams>,
+) -> Boolean<F>{
+    let (x, y) = point;
+
+    let mut x = x.clone();
+    let mut y = y.clone();
+
+    let b = BN256Fq::from_str(B).unwrap();
+    let mut b = BN256BaseNNField::allocated_constant(cs, b, params);
+
+    let mut x_squared = x.square(cs);
+    let mut x_cubed = x_squared.mul(cs, &mut x);
+
+    let mut x_cubed_plus_b = x_cubed.add(cs, &mut b);
+    let mut y_squared = y.double(cs);
+
+    BN256BaseNNField::equals(cs, &mut y_squared, &mut x_cubed_plus_b)
+}
+
+/// Checks that the passed point is on G2 `BN256` curve.
+/// The `Infinity` point is not counted as on curve.
+/// See https://hackmd.io/@jpw/bn254#Twists for further details.
+pub(crate) fn is_on_twist_curve<F: SmallField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    point: (&BN256Fq2NNField<F>, &BN256Fq2NNField<F>),
+    params: &Arc<BN256BaseNNFieldParams>,
+) -> Boolean<F>{
+    let (x, y) = point;
+
+    let mut x = x.clone();
+    let mut y = y.clone();
+
+    let b_c0 = BN256Fq::from_str(B_TWIST_C0).unwrap();
+    let b_c1 = BN256Fq::from_str(B_TWIST_C1).unwrap();
+
+    let b_c0 = BN256BaseNNField::allocated_constant(cs, b_c0, params);
+    let b_c1 = BN256BaseNNField::allocated_constant(cs, b_c1, params);
+
+    let mut b = BN256Fq2NNField::new(b_c0, b_c1);
+
+    let mut x_squared = x.square(cs);
+    let mut x_cubed = x_squared.mul(cs, &mut x);
+
+    let mut x_cubed_plus_b = x_cubed.add(cs, &mut b);
+    let mut y_squared = y.double(cs);
+
+    y_squared.equals(cs, &mut x_cubed_plus_b)
+}
+
+/// Check whether passed point is classified as `Infinity`.
+/// See https://eips.ethereum.org/EIPS/eip-196 for further details.
+// We use `UInt256` instead of `BN256BaseNNField`
+// because we need to be able to check the unmasked value.
+pub(crate) fn is_affine_infinity<F: SmallField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    point: (&UInt256<F>, &UInt256<F>),
+) -> Boolean<F> {
+    let (mut x, mut y) = point;
+    let x_is_zero = x.is_zero(cs);
+    let y_is_zero = y.is_zero(cs);
+
+    Boolean::multi_and(cs, &[x_is_zero, y_is_zero])
+}
+
+/// Check whether passed point in G2 is classified as `Infinity`.
+/// See https://eips.ethereum.org/EIPS/eip-196 for further details.
+// We use `UInt256` instead of `BN256BaseNNField`
+// because we need to be able to check the unmasked value.
+pub(crate) fn is_twist_affine_infinity<F: SmallField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    point: (&UInt256<F>, &UInt256<F>, &UInt256<F>, &UInt256<F>),
+) -> Boolean<F> {
+    let (mut x_c0, mut x_c1, mut y_c0, mut y_c1) = point;
+
+    let x_c0_is_zero = x_c0.is_zero(cs);
+    let x_c1_is_zero = x_c1.is_zero(cs);
+    let x_is_zero = x_c0_is_zero.and(cs, x_c1_is_zero);
```

Comment:
i think the single `and` calls can be removed and all 4 booleans can be put in the multi and gate, it should be a touch cleaner